### PR TITLE
*: support fixed dimension vector

### PR DESCRIPTION
**Pull request by EricZequan**
### What problem does this PR solve?

Issue Number: ref #54245 

Problem Summary:

### What changed and how does it work?
Support fixed size vector types like `VECTOR(3)`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
```mysql
mysql> DROP TABLE t;
Query OK, 0 rows affected (0.12 sec)

mysql> create table t(embedding VECTOR);
Query OK, 0 rows affected (0.06 sec)

mysql> INSERT INTO t VALUES ('[1,2,3]'), ('[4,5]'), ('[6]');
Query OK, 3 rows affected (0.01 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> ALTER TABLE t MODIFY COLUMN embedding VECTOR(0);
ERROR 1105 (HY000): vector has 3 dimensions, does not fit VECTOR(0)
mysql> ALTER TABLE t MODIFY COLUMN embedding VECTOR(16001);
ERROR 1105 (HY000): vector cannot have more than 16000 dimensions
mysql> ALTER TABLE t MODIFY COLUMN embedding VECTOR(3);
ERROR 1105 (HY000): vector has 2 dimensions, does not fit VECTOR(3)
mysql> DELETE FROM t WHERE VEC_DIMS(embedding) != 3;
Query OK, 2 rows affected (0.01 sec)

mysql> ALTER TABLE t MODIFY COLUMN embedding VECTOR(3);
Query OK, 0 rows affected (0.14 sec)

mysql> SHOW COLUMNS FROM t;
+-----------+------------------+------+------+---------+-------+
| Field     | Type             | Null | Key  | Default | Extra |
+-----------+------------------+------+------+---------+-------+
| embedding | vector<float>(3) | YES  |      | NULL    |       |
+-----------+------------------+------+------+---------+-------+
1 row in set (0.00 sec)

mysql> ALTER TABLE t MODIFY COLUMN embedding VECTOR;
Query OK, 0 rows affected (0.07 sec)

mysql> 
mysql> SHOW COLUMNS FROM t;
+-----------+---------------+------+------+---------+-------+
| Field     | Type          | Null | Key  | Default | Extra |
+-----------+---------------+------+------+---------+-------+
| embedding | vector<float> | YES  |      | NULL    |       |
+-----------+---------------+------+------+---------+-------+
1 row in set (0.00 sec)
```
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by breezewish — COMMENTED** [review]
image

This part is missing?

File: pkg/expression/integration_test/integration_test.go [hunk 1]
```diff
@@ -57,6 +57,147 @@ import (
...
+
+	// Fixed dimension to mixed dimension.
+	tk.MustExec("alter table t modify column embedding VECTOR")
+	tk.MustExec("insert into t values ('[1,2,3,4]')")
+
+	// Vector dimension MUST be equal or less than 16000.
+	tk.MustGetErrMsg("alter table t modify column embedding VECTOR(16001)", "vector cannot have more than 16000 dimensions")
+}
+
+func TestVectorVariable(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (EricZequan) [reply pkg/expression/integration_test/integration_test.go]: yep. Some test fail cuz the missing part. Now it is fixed. And the `TestVectorVariable` has removed.

File: pkg/types/datum.go [hunk 2]
```diff
@@ -927,7 +927,7 @@ func (d *Datum) compareMysqlTime(ctx Context, time Time) (int, error) {
 	}
 }
 
-func (d *Datum) compareVectorFloat32(sc Context, vec VectorFloat32) (int, error) {
+func (d *Datum) compareVectorFloat32(ctx Context, vec VectorFloat32) (int, error) {
```
[the remark this answers is not in the document]
Comment (EricZequan) [reply pkg/types/datum.go:930]: Other `compareXXX` function's parameters Name is ctx, just `compareVectorFloat32` is `sc`. So I rename it in order to keeping same.

[the remark this answers is not in the document]
Comment (breezewish) [reply pkg/types/datum.go:930]: Got it, nice change!

**Review by breezewish — APPROVED** [review]

File: pkg/types/vector.go [hunk 3]
```diff
@@ -55,6 +56,28 @@ func InitVectorFloat32(dims int) VectorFloat32 {
 	return VectorFloat32{data: data}
 }
 
+// CheckVectorDimValid checks if the vector's dimension is valid.
+func CheckVectorDimValid(dim int) error {
+	const (
+		maxVectorDimension = 16000
```
Comment (JaySon-Huang) [inline pkg/types/vector.go:62]: the maximum dimension supported in MySQL is 16383. We should align to it according to the feature spec. /cc @breezewish

Comment (breezewish) [reply pkg/types/vector.go:62]: Nice catch. Please also add a TiDB-CSE PR to update this limit. @EricZequan

Comment (EricZequan) [reply pkg/types/vector.go:62]: Roger that.🫡

**Review by tangenta — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**